### PR TITLE
Optimization using compile-time initialization and validation

### PR DESCRIPTION
**Pull request by gfraiteur**
Phil,

Thank you for writing about PostSharp. I hope you like it. It took me 8 years to build :)

Please consider these changes. They result in much faster execution at runtime because all reflection is done at build time. The PostSharp API is designed to be easy to get started with, but then users can go further if they want to do things optimally. The "tricks" I've used are documented on http://doc.sharpcrafters.com/postsharp-2.1/Content.aspx/PostSharp-2.1.chm/html/7480ca54-61c0-46c5-9914-60a58c3033e8.htm.

Don't hesitate to contact me at gael@fraiteur.net if you have any question.

-gael

### CODE REVIEW

File: NullGuard.PostSharp/EnsureNonNullAspect.cs [hunk 1]
```diff
@@ -20,86 +27,136 @@ public EnsureNonNullAspect(ValidationFlags validationFlags)
...
+                    }
+                }
             }
-            throw new ArgumentNullException(invalidArgument.Name);
+
+     
+
         }
 
         public override void OnExit(MethodExecutionArgs args)
```
Comment (gfraiteur) [inline NullGuard.PostSharp/EnsureNonNullAspect.cs]: OnEntry should be changed to OnSuccess because no validation should be done upon exception.

**Comment by haacked** [comment]
Hi @gfraiteur. So far, despite my nitpicking, I think PostSharp is pretty cool! And thanks for submitting a pull request! 

However, I ran my unit tests and they fail. Are you sure this code keeps the existing behavior? I added a `build.bat` (it uses Powershell to call `Build-Solution.ps1`) that runs all the unit tests (in case you don't have xUnit installed). You'll just need to rebase your branch against my master.

**Comment by gfraiteur** [comment]
I cared to respect all existing semantics and every test was green when I committed. 
After merging the your last commit, `build.bat` just runs fine. I also pushed another bug fixed (OnExit was used instead of OnSuccess, which was bad because contracts should not be checked upon exception).

**Comment by haacked** [comment]
This is great. Thanks!

**Pull request merged**